src/test/ui/rfc-2632-const-trait-impl/specialization/non-const-default-const-specialized.rs

```diff
@@ -0,0 +1,34 @@
+// Tests that a non-const default impl can be specialized by a const trait impl,
+// but that the default impl cannot be used in a const context.
```

Comment:
@lcnr And would you say that this should be disallowed too?

Reply:
Don't think so. I think this goes like this:

|| General impl, not const  | General impl, const |
|---| ------------- | ------------- |
| More specific impl, not const | Currently allowed  | Currently disallowed, should be disallowed  |
| More specific impl, const  | Currently disallowed, should be allowed  | Currently disallowed, should be allowed |

Reply:
I don't get why top right is any worse than bottom left.

Reply:
The same reason why you can have

```rust
trait Super {}
trait Sub : Super {}
trait A {
    fn foo<T: Sub>();
}
impl A for () {
    fn foo<T: Super>();
}
```
but not the other way around.

Reply:
Whoa I had no idea that was allowed. But it doesn't seem equivalent to me. When testing it out, `<() as A>::foo` still requires that `T: Sub` even though the impl has the less strict `T: Super` (you can even leave out the bound entirely on the impl and it still requires that `T: Sub`). [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=24f05ef3968676a7b6b0b30834bf6c65)

By your analogy, I take it that the non-const default impl is the more restrictive `T: Sub`, and the const specializing impl is the more permissive `T: Super`. But if the analogy holds, then I'd expect us to disallow the use of the const specializing impl in a const context.

Allowing bottom left and allowing top right feel the same to me because both of them are allowing you to do something with some of the impls but not with the others.

Reply:
The fact that we ignore the weaker bounds on the impl is something we want to change. This is https://github.com/rust-lang/rust/issues/100706 which currently wants to require an explicit `#[refine]` attribute to use the less strict associated items.

yeah, this table is the behavior i expected:

|| General impl, not const  | General impl, const |
|---| ------------- | ------------- |
| More specific impl, not const | Currently allowed  | should be disallowed  |
| More specific impl, const  | should be allowed  | should be allowed |

You could think of `const` functions as having the `trait NonConst: Const {}` as a bound. A function which is not const has more requirements for the caller (as they are not allowed to use this function in a const context).

The intent is not "doing something with some of the impls but not with the others". The intent is "loosening the restrictions on assoc items in specialized impls is ok". If you have some specialized impl, that impl should be usable in all ways the default impl can be used, and maybe more (by loosening a restriction). You should not loose capabilities by going from a default to a specialized impl.

does that explanation make sense to you? it feels somewhat difficult to explain :grin::sweat_smile:

Reply:
> does that explanation make sense to you? it feels somewhat difficult to explain

Yes, that explanation helps. Thanks.

Reply:
Any suggestion on what to put in the diagnostic when we disallow top right?

Reply:
Perhaps "one impl is more general"?